nixos/nftables: Allow use with iptables

jess-sch opened the pull request:
This pull request removes the checkScript portion of the nftables module (which currently prevents nftables from being used together with iptables) and updates the relevant documentation.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Since iptables and nftables do not actually conflict with each other, there's no real reason to artificially prevent people from combining them. What happens if you combine the two is [well-defined](https://wiki.nftables.org/wiki-nftables/index.php/Troubleshooting#Question_4._How_do_nftables_and_iptables_interact_when_used_on_the_same_system.3F).

The current "prevent nftables from being used with iptables" policy causes issues like #88643.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jess-sch committed:
nixos/nftables: Allow use with iptables

Since iptables and nftables do not actually conflict with each other, there's no real reason to artificially prevent people from combining them.
In fact, this practice is known to cause issues like #88643, which is fixed by this commit.

jess-sch committed:
nixos/nftables: Remove whitespace in line 21

Make the EditorConfig check pass

andir commented:
@ofborg test lxd-nftables

andir reviewed (changes requested):
I am not really happy with removing the warning.

Given the current state of nftables in NixOS we shouldn't allow enabling both nftables and iptables at the same time by accident. I could imagine that there are a few users who might assume that `nftables` (once enabled) is used for all the builtin firewalling and that `iptables` is no longer used.

Perhaps we need to be a bit more picky about the users intend. How about introducing an enum of sorts that lets you select if you want `iptables`, `nftables` or both?

In any way this requires a changelog entry to inform users about the change.

Another thing to keep in mind with this change: We have modules that operate under the assumption that only one of them is enabled but not both of them. Some of these seem to be the `sshguard` and the `fail2ban` modules.

jess-sch commented:
@andir would it be okay to just remove the checkScript (even if the assertion stays)? That way, the user would still be picking one firewall only, but at least nftables wouldn't randomly fail to update just because some application thought it would be a good idea to load iptables.

jess-sch committed:
nixos/nftables: Undo removal of assertion

Reduces #121517 to a simple removal of the runtime check, so that nftables and firewall can not be enabled together, but if they are both loaded, nftables will still be able to reload its configuration.

jess-sch committed:
nixos/nftables: Remove whitespace in empty lines

... to make EditorConfig happy

antifuchs reviewed (approved):
This PR fixes a pretty annoying problem for me (can't adjust firewall rules after tailscale comes up). I think @andir raises a valid point insofar as the wider nixos configuration ecosystem makes some baked-in assumptions about the type of firewall that's in place... but it really shouldn't be up to this PR to fix all of them, IMHO.

mweinelt commented:
If we migrated iptables to the iptables-nft wrapper, could we get rid of the warning because we'd be using nftables either way?

misuzu commented:
https://github.com/NixOS/nixpkgs/pull/81172 is now in nixos-unstable.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/moving-from-iptables-to-nf-tables/16396/2

mogorman commented:
is there anything blocking this being merged? or are @andir comments blocking?

Mic92 commented:
> is there anything blocking this being merged? or are @andir comments blocking?

They should not really apply anymore. It's the nftables backend in any case.
